ale_linters/cpp/clangtidy.vim

```diff
@@ -35,17 +43,45 @@ endfunction
 function! ale_linters#cpp#clangtidy#GetCommand(buffer) abort
     let l:checks = join(ale#Var(a:buffer, 'cpp_clangtidy_checks'), ',')
     let l:build_dir = s:GetBuildDirectory(a:buffer)
-
     " Get the extra options if we couldn't find a build directory.
     let l:options = empty(l:build_dir)
-    \   ? ale#Var(a:buffer, 'cpp_clangtidy_options')
-    \   : ''
-
-    return ale#Escape(ale_linters#cpp#clangtidy#GetExecutable(a:buffer))
-    \   . (!empty(l:checks) ? ' -checks=' . ale#Escape(l:checks) : '')
-    \   . ' %s'
-    \   . (!empty(l:build_dir) ? ' -p ' . ale#Escape(l:build_dir) : '')
-    \   . (!empty(l:options) ? ' -- ' . l:options : '')
+                \   ? ale#Var(a:buffer, 'cpp_clangtidy_options')
+                \   : ''
+
+    if !ale_linters#cpp#clangtidy#IsHeader(a:buffer)
+        return ale#Escape(ale_linters#cpp#clangtidy#GetExecutable(a:buffer))
+        \   . (!empty(l:checks) ? ' -checks=' . ale#Escape(l:checks) : '')
+        \   . ' %s'
+        \   . (!empty(l:build_dir) ? ' -p ' . ale#Escape(l:build_dir) : '')
+        \   . (!empty(l:options) ? ' -- ' . l:options : '')
+    else
+        let l:header_sourcefile = ale#Var(a:buffer, 'cpp_clangtidy_header_sourcefile')
+        " If no specific header options provided by user:
+        " Don't include build directory for header files, as compile_commands.json
+        " files don't consider headers to be translation units, and provide no
+        " commands for compiling header files.
+        " Use the same command as non-header files.
+        if empty(l:header_sourcefile)
+            let l:build_dir = ''
+            let l:options = ale#Var(a:buffer, 'cpp_clangtidy_options')
+            return ale#Escape(ale_linters#cpp#clangtidy#GetExecutable(a:buffer))
+            \   . (!empty(l:checks) ? ' -checks=' . ale#Escape(l:checks) : '')
+            \   . ' %s'
+            \   . (!empty(l:build_dir) ? ' -p ' . ale#Escape(l:build_dir) : '')
+            \   . (!empty(l:options) ? ' -- ' . l:options : '')
+        " If header_sourcefile: Don't use the current buffer as input! The user has to
+        " provide a source file (.cpp) to run the checker on it when editing
+        " the header.
+        " The build_dir is kept if present, the provided source file must be
+        " in the database.
+        else
+            return ale#Escape(ale_linters#cpp#clangtidy#GetExecutable(a:buffer))
+            \   . (!empty(l:checks) ? ' -checks=' . ale#Escape(l:checks) : '')
+            \   . (!empty(l:header_sourcefile) ? ' ' . ale#Escape(l:header_sourcefile) : '')
```

Comment:
You could just write `'  ' . (!empty(l:header_sourcefile) ? ale#Escape(l:header_sourcefile) : '%s')` with the way the code was structured before.